packages/protocol/contracts/verifiers/compose/ComposeVerifier.sol

```diff
@@ -0,0 +1,88 @@
+// SPDX-License-Identifier: MIT
+pragma solidity 0.8.24;
+
+import "../../common/EssentialContract.sol";
+import "../IVerifier.sol";
+
+/// @title ComposeVerifier
+/// @notice This contract is an abstract verifier that composes multiple sub-verifiers to validate
+/// proofs.
+/// It ensures that a set of sub-proofs are verified by their respective verifiers before
+/// considering the overall proof as valid.
+/// @custom:security-contact security@taiko.xyz
+abstract contract ComposeVerifier is EssentialContract, IVerifier {
+    struct SubProof {
+        address verifier;
+        bytes proof;
+    }
+
+    event InvalidSubProof(address indexed verifier, bytes returnData);
+
+    error INVALID_VERIFIER();
+    error INSUFFICIENT_PROOF();
+
+    /// @notice Initializes the contract.
+    /// @param _owner The owner of this contract. msg.sender will be used if this value is zero.
+    /// @param _rollupAddressManager The address of the {AddressManager} contract.
+    function init(address _owner, address _rollupAddressManager) external initializer {
+        __Essential_init(_owner, _rollupAddressManager);
+    }
+
+    /// @notice Verifies one or more sub-proofs.
+    /// @param _ctx The context of the proof verification.
+    /// @param _tran The transition to verify.
+    /// @param _proof The proof to verify.
+    function verifyProof(
+        Context calldata _ctx,
+        TaikoData.Transition calldata _tran,
+        TaikoData.TierProof calldata _proof
+    )
+        external
+    {
+        (address[] memory verifiers, uint256 threshold) = getSubVerifiersAndThreshold();
+        SubProof[] memory subproofs = abi.decode(_proof.data, (SubProof[]));
+        uint256 numVerified;
+
+        for (uint256 i; i < subproofs.length; ++i) {
+            if (subproofs[i].verifier == address(0)) revert INVALID_VERIFIER();
+
+            // find the verifier
+            bool verifierFound;
+            for (uint256 j; j < verifiers.length; ++j) {
+                if (verifiers[j] == subproofs[i].verifier) {
+                    verifierFound = true;
+                    verifiers[j] = address(0);
+                }
+            }
+
+            if (!verifierFound) revert INVALID_VERIFIER();
+
+            (bool success, bytes memory returnData) = subproofs[i].verifier.call(
+                abi.encodeCall(
+                    IVerifier.verifyProof,
```

Comment:
Calling `SgxVerifier`'s `verifyProof()` - if we will start using the TEE any, we need to modify the code because SgxVerifier.sol line 146, it has a guard:
`onlyFromNamed(LibStrings.B_TAIKO)`.

Maybe best to mention it as a todo.

Reply:
It was initially added by Dani for security concern I guess @adaki2004? And yes, I think it can be removed now.

Reply:
Im not sure if we can / shall remove it.

From mobile ATM but there was an issue allowing calling the contract by anybody.

I think it was the replay issue. So if caller is not the TaikoL1 and the call succeeds, it already exchanges the keys, so that the call from TaikoL1 womt suceed, this is why we need here. (If not mistaken).

Reply:
yes, but why a prover does that? it seems no harm for the others as he can not update others key, only negative impact to himself?

Reply:
I read the code again, it seems we can indeed remove the check. Without the check, basically if someone has access to an instance's private-key, he can then remove the instance (and potentially add a new one). But this will not affect others as this instance will never be used by others.

Reply:
> yes, but why a prover does that? it seems no harm for the others as he can not update others key, only negative impact to himself?

No, (afaik) it is not about that.

Imagine Bob is an honest (sgx) prover. Alice is a malicious actor who wants to trick Bob into losing money.
Alice monitors the mempool and when she sees the proving transaction, she extracts the proof and submit directly to the SgxVerifier contract with higher gas price.
Bob’s proof then would be revertes and possibly would loose money (bond). This guard is to prevent that situation.
(Basically it would work all the time and can increase proving time so that causes harm to alk the SGX prover to pay up to “infonity” just to prove a block and get back a bond).

it was also flagged  by SigmaPrime audit.

Reply:
> I read the code again, it seems we can indeed remove the check. Without the check, basically if someone has access to an instance's private-key, he can then remove the instance (and potentially add a new one). But this will not affect others as this instance will never be used by others.

I'm referring to this audit finding.
<img width="971" alt="kép" src="https://github.com/user-attachments/assets/eb573c5c-ae64-4771-9aeb-bedb235b0038">

Reply:
Can Alice frontruns Bob it means Bob shared his private key with Alice so they are considered the same party, right? Otherwise, how can Alice create a signature for Bob's instance?


```
      address oldInstance = ECDSA.recover(
            LibPublicInput.hashPublicInputs(
                _tran, address(this), newInstance, _ctx.prover, _ctx.metaHash, taikoChainId()
            ),
            _proof.data[24:]
        );
```

Reply:
> Can Alice frontruns Bob it means Bob shared his private key with Alice so they are considered the same party, right? Otherwise, how can Alice create a signature for Bob's instance?
> 
> ```
>       address oldInstance = ECDSA.recover(
>             LibPublicInput.hashPublicInputs(
>                 _tran, address(this), newInstance, _ctx.prover, _ctx.metaHash, taikoChainId()
>             ),
>             _proof.data[24:]
>         );
> ```
Maybe i'm wrong or i dont see if we changed something in between, but here is my idea:

1. Bob has to re-create this "payload" off-chain, right ? (to sign it)
```
LibPublicInput.hashPublicInputs(
                _tran, address(this), newInstance, _ctx.prover, _ctx.metaHash, taikoChainId()
            ),
```
            
2. Alice cannot have access to the private key of Bob BUT she **sees** Bob's transaction **in the mempool**. She can extract this information then submit the smart contract call with higher gas tip. (Without that guard).

3. So Alice take advantage of the fact she can replace Bob's with the new instance, so that Bob cannot verify the transition with the old anymore.

Imo this creates the security vulnerability. (Please let me know if it is not or we mitigated somehow.. But i still think we need the `onlyFromNamed` ).

Reply:
Because of the potential key switching it is indeed a problem and should be guarded.

Reply:
> 2\. Alice cannot have access to the private key of Bob BUT she **sees** Bob's transaction **in the mempool**. She can extract this information then submit the smart contract call directly to SgxVerifier, with higher gas tip. (Without that guard).

I see.